SelectMenu 컴포넌트 제작

wooojini opened the pull request:
## 📑 제목

#35 SelectMenu 컴포넌트 제작


## 💬 작업 내용

> 구현 내용 및 작업 했던 내역

- [x] **SelectMenu 컴포넌트 제작**
   - 기존 select, option 태그의 스타일 커스텀 한계로 인해 material-ui popover 컴포넌트 활용 제작
![ezgif com-gif-maker (33)](https://user-images.githubusercontent.com/32856129/114226729-ec22e080-99ae-11eb-9e62-ee8767128540.gif)
- [x] **TimeSelectMenu 컴포넌트 제작**
    - 기존 input type='time'의 경우, 스타일 커스텀에 한계가 존재해서 popover 컴포넌트 활용하여 제작
    - UX를 고려하여 사용자가 시간 메뉴를 선택했을 때,  메뉴창이 닫히지 않고 열린 상태로 유지되고 메뉴 바깥 쪽을 클릭했을 때 메뉴가 닫히도록 제작 
![ezgif com-gif-maker (34)](https://user-images.githubusercontent.com/32856129/114227316-ce09b000-99af-11eb-8c2d-3081a2345a63.gif)
- [x] **LectureSearchFilter 컴포넌트 제작**
![ezgif com-gif-maker (35)](https://user-images.githubusercontent.com/32856129/114228230-183f6100-99b1-11eb-822f-4de24f033e06.gif)
- [x] **TimeTableAddForm 컴포넌트 제작**
![ezgif com-gif-maker (36)](https://user-images.githubusercontent.com/32856129/114228817-f7c3d680-99b1-11eb-815c-015783c7709c.gif)
- [x] 기타 작업사항
   - start~end 사이의 수를 offset만큼 건너뛰어서 반복해주는 이터레이터 util 함수를 작성했습니다.
      - for문을 이용해서 반복해야만하는 상황에서 사용해주세요!
      ```
         //1~10 수를 콘솔에 출력하는 예시 코드
         for(let i = 1; i <= 10; i++) console.log(i); //기존 for 문

         for(num of range(1,10)) console.log(i) //for..of문 활용
         Array.from(range(1,10)).forEach((num) => console.log(num)); //배열함수 활용
      ```
   - 불필요한 eslint 옵션 off
      -  ++연산자를 제한하는 옵션을 off 했습니다. 추후에 issue를 새로 열어서 현재 eslint 설정 파일이 아닌 컴포넌트 코드 파일에 추가되어있는 옵션 off하는 코드들을 eslint 설정파일에서 off 하도록 해야할 듯 합니다!
     ```
        /* eslint-disable react/jsx-closing-bracket-location */   // 불필요한 옵션 때문에 컴포넌트 파일에 요런 코드가 계속 추가됨.
        import React from 'react';
         ...(생략)....
     ```


## 🚧 PR 특이 사항

> PR을 볼 때 주의깊게 봐야하거나 말하고 싶은 점

- 이전에 진혁님께서 이야기해주신데로 모바일 대응을 위해 추후에 select, option 태그로 바꿔주는 로직이 추가되어야 합니다.
- 강의 필터링이나 나만의 시간표 추가 비즈니스 로직 코드는 작업하지 않고, 뷰 작업만 완료한 상태입니다. 추후에 서버 API와 확실하게 연동 후, 클라이언트 스토어 구조도 fix 되면 작업합시다!
- **zepline 디자인과 약간 다른사항이 존재합니다.**
  - zepline에서는 아래와 같이 컴포넌트 드랍메뉴 간에 약간의 간격이 존재합니다.
 <img src='https://user-images.githubusercontent.com/32856129/114229099-59844080-99b2-11eb-99e5-5ad53ef80fb4.png' width=300/>

  - 실제 제작한 SelectMenu, TimeSelectMenu 컴포넌트는 클릭했을 때 보여지는 메뉴와의 간격이 딱 붙은 상태로 보여집니다.
     - popover 컴포넌트를 활용했는데, popover 컴포넌트는 모달처럼 전체 배경을 깔고, 메뉴를 fixed로 화면에서 좌표를 찾아서 위치를 잡습니다! 제공해주는 props를 이용하면 위치를 잡을 기준 컴포넌트를 지정해서 위치를 잡을 수 있는데, 위 디자인과 같이 간격을 주려면 위치를 좌표로 설정해야해서 계산 로직이 추가로 작성되어야합니다.
     - https://material-ui.com/components/popover/#anchor-playground
     -  (위 링크를 참고하셔서 조작해보시면 이해되실 겁니다!)
     - 디자인 상으로 미미한 부분이라 불필요하게 로직이 추가되는 것 같아서 일단 보류했는데 진혁님 의견이 궁금하네용
   
  - **Popover 컴포넌트가 클릭 당시 브라우저 화면 height 사이즈에 해당하는 배경을 깔고, 메뉴의 position을 fixed로 좌표를 지정합니다! 문제는 브라우저 화면 사이즈가 메뉴가 보여질 만큼 충분히 있지 않으면 메뉴 좌표가 그 만큼 상단으로 밀리게 됩니다!**
     - 아래와 같이 브라우저 배율을 조정해서 메뉴가 보여질 충분한 height가 확보되면 정상적으로 메뉴가 위치를 잡습니다!
     - 제 생각은 한표에 min-height를 설정해서 기본 height를 지정해주는게 어떨까 합니다! 기본으로 여유롭게 지정해주면 항상 스크롤도 생겨서 스크롤 때문에 UI가 움직이는 현상도 해결될 듯 합니다!
 
![ezgif com-gif-maker (37)](https://user-images.githubusercontent.com/32856129/114230601-4ffbd800-99b4-11eb-8780-dabafe82d4c0.gif)

- 생각보니 한표 애플리케이션에 footer가 없네용???! min-height 지정하면서 footer도 추가해서 한표 관련 정보 보여주어도 좋을 듯 합니다!

wooojini committed:
test: 스토리북 css reset 설정 추가

- 스토리북 테스팅시 스타일을 reset하고 개발할 수 있도록 reset.css import
  구문 추가

wooojini committed:
feat: SelectMenu 컴포넌트 제작

- SelectMenu 컴포넌트 제작 및 스토리 작성

wooojini committed:
test: 불필요한 테스트 파일 삭제

- 프로젝트 초기 테스트 환경 구성을 위한 샘플 테스트 코드 삭제

wooojini committed:
feat: range 유틸 함수 작성

- start~end 사이의 수를 offset 간격으로 반복하는 이터레이터 유틸 함수
- range 유틸 테스트 코드 작성

wooojini committed:
chore: eslint ++연산자 제한 옵션 off

- ++연산자 사용을 제한하는 옵션 off

wooojini committed:
style: Button 컴포넌트 스타일 수정

- Add Button px단위 rem단위로 수정

wooojini committed:
refactor: SelectMenu 컴포넌트 디자인 수정

- SelectMenu 컴포넌트 Popover 컴포넌트를 이용해서 직접 제작
- 외부에서 width값을 설정해서 크기를 조절하도록 수정 및 스토리북 코드
  반영

wooojini committed:
feat: TimeSelectMenu 컴포넌트 제작

- 시간 선택을 위한 TimeSelectMenu 컴포넌트 제작
- 기존 input type=time으로 제공되는 UI는 브라우저마다 다르고, 스타일
  커스텀에 한계가 존재해서 별도로 제작
- TimeSelectMenu 컴포넌트 스토리 작성

wooojini committed:
feat: LectureSearchFilterMenu 컴포넌트 제작

- 강의 검색 필터를 위한 LectureSearchFilterMenu 컴포넌트 제작
- LectureSearchFilterMenu 스토리 작성

wooojini committed:
feat: TimeTableAddForm 컴포넌트 제작

- 나만의 시간표 추가를 위한 TimeTableAddFrom 컴포넌트 제작
- TimeTableAddForm 스토리 작성

wooojini committed:
style: 메인페이지 레이아웃 수정

- LectureSearchFilterMenu, TimeTableAddForm 컴포넌트 메인페이지 레이아웃
  추가

jinhyukoo commented on client/.eslintrc.js:
```diff
@@ -44,5 +44,6 @@ module.exports = {
     'no-shadow': 'off',
     '@typescript-eslint/no-shadow': 'error',
+    'no-plusplus': 'off',
   },
```
ㅋㅋㅋㅋ 예전부터 뭔가 약간(?)씩 거슬리던 부분이었는데 잘 추가해주신 것 같아요 ㅎㅎ

jinhyukoo commented on client/.storybook/preview.js:
```diff
@@ -1,4 +1,5 @@
 import React from 'react';
+import '@/index.css';
 import { ThemeProvider } from '@material-ui/core/styles';
```
👍

jinhyukoo commented on client/src/components/UI/atoms/Button/Button.tsx:
```diff
@@ -44,11 +44,12 @@ const useStyles = makeStyles({
-    width: '39.6px',
-    height: '15.5px',
+    width: '3.75rem',
+    height: '2.015625rem',
```
지난번에도 궁금했는데 소수점 6번째 자리까지 나오는 수치는 어떻게 나오는건가용...!?

wooojini replied:
@jinhyukoo px단위로 먼저 작업을 하고, 16px(1rem)로 나누어서 rem 단위로 적용했어용!

jinhyukoo commented on client/src/components/UI/atoms/SelectMenu/SelectMenu.tsx:
```diff
@@ -0,0 +1,148 @@
+/* eslint-disable jsx-a11y/no-noninteractive-element-interactions */
+/* eslint-disable jsx-a11y/no-static-element-interactions */
+/* eslint-disable jsx-a11y/click-events-have-key-events */
+/* eslint-disable react/jsx-closing-bracket-location */
```
나중에 중복되는 lint option들은 config쪽에 정의를 해버리면 더 좋겠네용

wooojini replied:
@jinhyukoo 맞습니다! 이슈로 등록했습니다 👍

jinhyukoo commented on client/src/components/UI/atoms/SelectMenu/SelectMenu.tsx:
```diff
@@ -0,0 +1,148 @@
+);
+
+const SelectMenu = ({ menuLabel, menus, dropMenuWidth = 'auto', onSelectMenuChange }: SelectMenuProps): JSX.Element => {
+  const [anchorEl, setAnchorEl] = React.useState<Element | null>(null);
```
useState가 import 되어 있는데 React.useState로 표기하신 다른 이유가 있을까요? 제가 이 부분을 정확히 몰라서 여쭤봅니당

wooojini replied:
@jinhyukoo 이건..... material ui 예제코드의 잔재입니다... 코드 머지 전에 수정할게용 ㅎㅎ

jinhyukoo commented on client/src/components/UI/atoms/SelectMenu/SelectMenu.tsx:
```diff
@@ -0,0 +1,148 @@
+  };
+
+  const onMenuClickListener = (event: React.MouseEvent<HTMLElement>) => {
+    const target = event.target as HTMLElement;
```
오 혹시 이 코드가 어떤 걸 의미하는지 여쭤봐도 될까요? 처음보는 문법이에용 ㅎㅎ

wooojini replied:
@jinhyukoo 타입단언입니다! 이벤트 리스너의 인자로 들어오는 이벤트 객체는 종류가 다양한 유니온 타입이라 event.target.dataset 이런식으로 접근해도 타입스크립트가 이벤트 객체로 어떤 타입이 들어오는지 추론을 하지 못해서 종종 에디터에서 오류가 많이 발생합니다! as 키워드를 사용하면 해당 타입이 들어온다는 것을 단언할 수 있습니다! 타입을 개발자가 단언하는 것이기 때문에 확실히 해당 타입이 들어오는 것을 인지하고나서 사용해야하지만 위의 경우에는 이벤트 리스너로 반드시 menuClick에만 사용하기 때문에 사용했습니다 ㅎㅎ

jinhyukoo replied:
아하 그렇군요! 좋은 정보 감사합니다!! 다음에 저도 사용해보고 싶네용 ㅎㅎ

jinhyukoo commented on client/src/components/UI/atoms/SelectMenu/SelectMenu.tsx:
```diff
@@ -0,0 +1,148 @@
+    if (!liElement) return;
+
+    const { dataset } = liElement;
+    setSelectValue(dataset?.title ?? '');
```
이것두요...!

wooojini replied:
@jinhyukoo 옵셔널 체이닝(optional chaining)과 널 병합 연산자(nullish coalescing operator)입니다. 

객체에서 접근하는 프로퍼티가 없으면 아래와 같이 에러가 발생하는 상황이 종종 발생합니다. 이러한 상황에 옵셔널 체이닝을 사용하면 유용합니다. 옵셔널 체이닝은 **?.'앞’의 평가 대상이 undefined나 null이면 평가를 멈추고 undefined를 반환합니다.**
```
let user = {}; // 주소 정보가 없는 사용자
alert(user.address.street); // TypeError: Cannot read property 'street' of undefined
```
 
널 병합 연산자는 **null이나 undefined인 변수 처리를 ??을 이용해서 처리할 수 있습니다.** 
```
a ?? b의 평가 결과는 다음과 같습니다.

a가 null도 아니고 undefined도 아니면 a
그 외의 경우는 b
```

이를 이용해서 위의 코드 같은 경우에는 dataset이 null이나 undefined가 아니면 title 프로퍼티를 참조하고, title 프로퍼티가 null or undefined 인 경우 ''로 초기값을 지정하고, 아니면 title 프로퍼티값을 selectValue값으로 지정하는 코드입니다!

jinhyukoo replied:
오오 이것도 굉장히 꿀팁이네요! 코드를 간결하게 할 때 유용하게 쓰일 것 같아요~ 감사합니다!!

jinhyukoo commented on client/src/components/UI/atoms/TimeSelectMenu/TimeSelectMenu.tsx:
```diff
@@ -0,0 +1,213 @@
+  const [isSelected, setIsSelected] = useState(false);
+  const [selectAMPM, setSelectAMPM] = useState('오전');
+  const [selectHour, setSelectHour] = useState('01');
+  const [selectMinute, setSelectMinute] = useState('00');
```
뜻으로 본다면 selected가 더 괜찮을 것 같기도 한데 중요한 이슈는 아니니 패스 하시죵 ㅎㅎ

wooojini replied:
@jinhyukoo 요것도 selected로 명확하게 수정 후에 머지할게용 ㅎㅎ

jinhyukoo commented on client/src/components/UI/molecules/LectureSearchFilter/LectureSearchFilter.tsx:
```diff
@@ -0,0 +1,70 @@
+      <Box className={classes.timeSelect}>
+        <TimeSelectMenu {...timeSelectMenu} dropMenuWidth="6.75rem" />
+      </Box>
+      <span>&#126;</span>
```
혹시 요건 뭔가용?

wooojini replied:
@jinhyukoo HTML 특수문자 숫자코드입니다! ~ 문자에 대한 숫자코드에요! 자세한 내용은 아래 링크를 방문하시면 바로 이해하실 수 있을겁니다!

https://tutorialpost.apptilus.com/code/posts/html/html-special-char/

jinhyukoo replied:
아하 저는 그냥 백틱을 사용해서 표현해 왔는데 이런 방법도 있군요! 감사해요!!

jinhyukoo commented on client/src/components/UI/molecules/TimeTableAddForm/TimeTableAddFormContent.tsx:
```diff
@@ -0,0 +1,91 @@
+      </Box>
+      <Box className={classes.timeSelect}>
+        <TimeSelectMenu {...timeSelectMenu} dropMenuWidth="6.75rem" />
+      </Box>
```
props로 전달되는 dropMenuWidth가 상위 컴포넌트의 width와 같은 값을 가지고 있는데 유지보수를 위해서라면 상수로 선언해주는 것도 괜찮겠다는 생각이 드네용! 하지만 겨우 이걸 위해서 변수가 늘어나는 것도 좀 애매한 것 같기도 하고... 뭐가 더 좋은 방법인지 잘 모르겠습니다 ㅋㅋㅋ

wooojini replied:
@jinhyukoo 진혁님과 똑같은 고민을 하다가 처리를 안한 부분인데..... 요기도 상수처리를 해서 머지할게용 ㅎㅎ

jinhyukoo replied:
진짜 뭐가 더 좋은 방법인지 모르겠네요 ㅋㅋㅋㅋ 사실 저도 이런 코드들을 많이 남겨놔서 리뷰하기가 망설여지긴 했는데... 고쳐서 머지시켜주신다면 거부하지는 않겠습니다 ㅋㅋㅋ

jinhyukoo reviewed (approved):
고생 많으셨습니다~ 약간의 디자인 변경사항은 크게 상관 없을 것 같아요! footer는 기획서에 없어서 생각 못하고 있었는데 하나 만들어야 겠네용 ㅋㅋㅋㅋㅋ 리뷰 내용들이 개선보다는 질문 위주라서 죄송함니당...ㅎ

wooojini committed:
refactor: SelectMenu, TimeSelectMenu 컴포넌트 코드 오타 수정

- React.useState를 useState로 수정
- TimeSelctMenu 컴포넌트 상태 변수를 selectXXX에서 selectedXXX로
  명확하게 네이밍 변경

wooojini committed:
refactor: LectureSearchFilter, TimeTableAddFormContent width 값 상수 처리

- 본문 내용 (여기에 작성)

wooojini committed:
style: Header가 스크롤 여부에 따라 움직이는 현상 제거

- 화면사이즈가 작은 경우 TimeSelectMenu가 보여질 공간이 충분치 않아서  min-height를 임시로 1000px로 설정
- Header 컴포넌트가 width 값을 가지지 않고, left, right 값으로 위치를
  조정하여 스크롤 여부에 따라 영향을 받는 문제를 width 값을 100%로 수정
  및 top 값에만 영향을 받도록 코드 수정

wooojini committed:
feat: scroll util 추가

- 브라우저의 스크롤을 조작하는 util 함수를 정의

wooojini committed:
fix: SelectMenu, TimeSelectMenu 컴포넌트 클릭시 메뉴의 위치가 올라가는 버그 수정

- 메인페이지 min-height 지정으로 인한 스크롤을 조작하여 메뉴 클릭시
  스크롤을 맨 밑으로 스크롤하여 메뉴창을 보여줄 height 확보 후, 메뉴창이
  열리도록 수정

wooojini committed:
fix: 헤더 스크롤 여부에 따른 padding 설정으로 인한 UI 움직이는 버그 수정

- SelectMenu, TimeSelectMenu 클릭 시 스크롤을 제거하고, padding-right
  값이 지정되어 UI가 움직이는 현상을 헤더에 padding-right값을 명시적으로
  지정하여 해결

wooojini commented:
@jinhyukoo 진혁님! 리뷰해주신 내용을 포함해서 몇가지 수정사항을 반영했습니다! 확인해주시고, 문제 없다면 머지할게용! ㅎㅎ

- React.useState(material-ui 예제코드 잔재)를 useState로 수정
- TimeSelectMenu 상태값 selectXXX에서 selectedXXX로 명확하게 네이밍 변경
- LectureSearchFilter, TimeTableAddFormContent 레이아웃 지정시 매직넘버들을 상수처리
- 메인페이지 min-height를 임시로 1000px로 지정
- SelectMenu, TimeSelectMenu 클릭시 메뉴창을 보여줄 height를 확보하기 위해 스크롤을 맨 밑으로 이동시키고, 메뉴창을 열도록 코드 수정
- 스크롤 기능은 자주 활용될 수 있어 별도의 scroll util 모듈로 정의
- 스크롤 여부에 따라 움직이는 현상을 제거하기 위해 헤더에 padding-right 값을 지정
   - 메뉴 창이 열리면 스크롤이 사라지고, 21px 만큼 padding-right를 material-ui에서 스타일을 자동으로 추가합니다
   - 헤더를 제외한 UI들은 padding 값이 지정되어서 움직이지 않는 것 처럼 보이지만, 헤더는 padding 스타일이 적용되지 않아 움직이는 것처럼 느껴집니다. 명시적으로 padding 값을 주어서 해결
   
![ezgif com-gif-maker (38)](https://user-images.githubusercontent.com/32856129/114372165-35944b00-9bbc-11eb-915f-4aeb39ca83ff.gif)
